httpclient/api_client.go

```diff
@@ -0,0 +1,386 @@
+	if cfg.Transport != nil {
+		return cfg.Transport
+	}
+	return &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		ForceAttemptHTTP2:     true,
+		MaxIdleConns:          100,
+		MaxIdleConnsPerHost:   runtime.GOMAXPROCS(0) + 1,
+		IdleConnTimeout:       180 * time.Second,
+		TLSHandshakeTimeout:   30 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: cfg.InsecureSkipVerify,
+		},
+	}
+}
+
+var DefaultClient = NewApiClient(ClientConfig{
+	ErrorRetriable:     DefaultErrorRetriable,
+	ErrorMapper:        DefaultErrorMapper,
+	HTTPTimeout:        30 * time.Second,
+	RetryTimeout:       5 * time.Minute,
+	RateLimitPerSecond: 30,
+})
+
+func NewApiClient(cfg ClientConfig) *ApiClient {
+	cfg.HTTPTimeout = time.Duration(orDefault(int(cfg.HTTPTimeout), int(30*time.Second)))
+	cfg.DebugTruncateBytes = orDefault(cfg.DebugTruncateBytes, 96)
+	cfg.RetryTimeout = time.Duration(orDefault(int(cfg.RetryTimeout), int(5*time.Minute)))
+	cfg.HTTPTimeout = time.Duration(orDefault(int(cfg.HTTPTimeout), int(30*time.Second)))
+	rateLimiter := rate.NewLimiter(rate.Limit(orDefault(cfg.RateLimitPerSecond, 15)), 1)
+	if cfg.ErrorMapper == nil {
+		// default generic error mapper
+		cfg.ErrorMapper = DefaultErrorMapper
+	}
+	if cfg.ErrorRetriable == nil {
+		// by default, we just retry on HTTP 429/504
+		cfg.ErrorRetriable = DefaultErrorRetriable
+	}
+	return &ApiClient{
+		config:      cfg,
+		rateLimiter: rateLimiter,
+		httpClient: &http.Client{
+			Timeout:   cfg.HTTPTimeout,
+			Transport: cfg.httpTransport(),
+		},
+	}
+}
+
+type ApiClient struct {
+	config      ClientConfig
+	rateLimiter *rate.Limiter
+	httpClient  *http.Client
+}
+
+type DoOption struct {
+	in   RequestVisitor
+	out  func(body *responseBody) error
+	body any
+}
+
+// Do sends an HTTP request against path.
+func (c *ApiClient) Do(ctx context.Context, method, path string, opts ...DoOption) error {
+	visitors := c.config.Visitors[:]
+	for _, o := range opts {
+		if o.in == nil {
+			continue
+		}
+		// merge client-wide and request-specific visitors
+		visitors = append(visitors, o.in)
+	}
+	var requestBody any
+	for _, o := range opts {
+		if o.body == nil {
+			continue
+		}
+		requestBody = o.body
+	}
+	responseBody, err := c.perform(ctx, method, path, requestBody, visitors...)
+	if err != nil {
+		return err
+	}
+	for _, o := range opts {
+		if o.out == nil {
+			continue
+		}
+		err = o.out(responseBody)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *ApiClient) fromResponse(r *http.Response) (responseBody, error) {
+	if r == nil {
+		return responseBody{}, fmt.Errorf("nil response")
+	}
+	if r.Request == nil {
+		return responseBody{}, fmt.Errorf("nil request")
+	}
+	streamResponse := r.Request.Header.Get("Accept") != "application/json" && r.Header.Get("Content-Type") != "application/json"
```

Comment:
We also need to add support for response headers but it could be done in separate PR since it's not immediately related to the changes in this PR and is more of a Feature Request.

Reply:
> We also need to add support for response headers

@tanmay-db it's there - `WithCaptureHeader()`

Reply:
> WithCaptureHeader and WithData

i'll rename `WithHeader` to `WithRequestHeader` and `WithCaptureHeader` to `WithResponseHeader`